### ISSUE TITLE
xterm write not work sometimes?

### ISSUE THREAD

**Issue by alanhg** [xtermjs/xterm.js#3910]
![image](https://user-images.githubusercontent.com/9245110/179788579-fc62e0d6-4e02-48cb-9ce3-47ca024a35d3.png)
![image](https://user-images.githubusercontent.com/9245110/179788673-68c0aa62-d3cf-4e05-b6ee-0fccab78658f.png)

the console is ok but xterm.write not render

write not throw error.

## Details
- Browser and browser version:  Chrome v103.0.5060.114
- OS version:  Monterey 12.4
- xterm.js version:  4.19.0

### Steps to reproduce

 https://github.com/alanhg/express-demo/blob/e458c584ff138450792cf4c38f6ad33ca191b678/views/xterm.ejs#L310-L310

**Comment by Tyriar** [comment]
It's not clear exactly what the issue is. You're paused on a breakpoint, rendering is done in the animation frame so it wouldn't show up until resumed. If you don't mean that you could investigate further, but this definitely works or no users of xterm.js would work.

**Comment by Tyriar** [comment]
It's because of the animation frame part, we update the model immediately and print it to the screen later. Printing after every write would be very slow

**Comment by alanhg** [comment]
Then now I have immediate printing needs, or I want to collect several writes and then print instead of the final overall print. What should I do?


For example, for the current demand, I want to print the upload progress, but now the effect is that I will wait for a while and directly print out 100%

**Comment by alanhg** [comment]
> 

My requirement is to print the progress to terminal

**Comment by Tyriar** [comment]
I don't mean to sound abrasive but this is a bug tracker and not a support forum. I'm sure there's nothing wrong with the library, beyond the help I gave above, you will need to figure this out on your own or find help elsewhere as we're quite busy.

**Comment by alanhg** [comment]
> I don't mean to sound abrasive but this is a bug tracker and not a support forum. I'm sure there's nothing wrong with the library, beyond the help I gave above, you will need to figure this out on your own or find help elsewhere as we're quite busy.
ok, thank you.

**Comment by xobs** [comment]
Thanks for responding to this issue with suggestions -- they were insightful.

I'm experiencing a similar issue with Xterm.js, though my situation is a bit different as I'm using it with React, and this is my first time using that framework. I'm using `xterm-for-react`, and I'm only able to write to the terminal when it first appears.

Curiously, `.writeSync()` works, but `.write()` does not, and the callback you mentioned never gets called.

My guess is that React does something funky with how xtermjs implements its main loop. Maybe they do something to inhibit `setTimeout()`, or maybe the buffer ends up empty. There are about three layers of indirection and minimization to cut through, and `xterm-for-react` doesn't appear to want to build on Windows since it really wants GTK for some reason. Time to build it in a Linux VM...

Either way, I appreciate your insignt into how things are set up, and it's been motivation to understand how these components all work together.

**Comment by Tyriar** [comment]
@xobs writeSync should generally not be used as it can break under certain circumstances. The async equivalent of this:

```ts
writeSync('abc');
// Check buffer
```

Is this:

```ts
term.write(data, () => {
  // Check buffer
});
```

writeSync also won't render to the screen synchronously, only to the backing buffer model.

> Either way, I appreciate your insignt into how things are set up, and it's been motivation to understand how these components all work together.

Summary of how it works:

- `write('abc', cb)` adds 'abc' to the write queue
- A `setTimeout(0)` is triggered to parse data of the write queue on the next frame, the amount of data parsed is approximately what can be done in 16ms
- Once `'abc'` is parsed, trigger `cb`
- `requestAnimationFrame` is triggered to render the updates to the screen

All these delayed frame/timeouts are done to keep the page responsive, even when hit with huge amounts of data as it would sit pending on the write queue

**Comment by xobs** [comment]
I moved `XTerm.tsx` from https://github.com/robert-harbison/xterm-for-react/blob/master/src/XTerm.tsx into my project, and rebuilt `xterm.js` without minification. This allowed me to step through and see what was going on.

I'm not sure if this is an issue or not, but the way React works is a function called `componentDidMount()` gets called when a component appears, and then `componentWillUnmount()` gets called when that function disappears.

It looks like what's happening is `xterm-for-react` is calling xterm's `.dispose()` function when the component gets unmounted, which destroys `Terminal.write()`. This means that any subsequent writes will be ignored. However, as I discovered, `Terminal._core.writeSync()` will continue to work.

The `xterm-for-react` then calls `Terminal.open()` afterwards to reinitialize the display, however this function does not reinstate the `.write()` call, so it continues to output no data.

I think there's an issue here in that the `.dispose()` documentation makes no mention of being a trapdoor function. It does seem from the API documentation that you ought to be able to run `.dispose()` followed by `.open()`.

I'll open a new issue related to this, since it's probably poor form to append comments to a [tangentially-related] bug that exhibits the same vague symptoms. Still, it was nice to get to the bottom of this. I enjoy how readable this project is.
